ICE: coverage: `attempt to add a 'Counter' to a BCB target with existing incoming edge counters`

<!--
[31mICE[0m: Rustc ./early_bound.rs '-Cinstrument-coverage -ooutputfile -Zdump-mir-dir=dir' 'thread 'rustc' panicked at compiler/rustc_mir_transform/src/coverage/counters.rs:125:9: 'attempt to add a `Counter` to a BCB target with existing incoming edge counters'', 'thread 'rustc' panicked at compiler/rustc_mir_transform/src/coverage/counters.rs:125:9: 'attempt to add a `Counter` to a BCB target with existing incoming edge counters''
File: /tmp/im/early_bound.rs
-->

auto-reduced (treereduce-rust):
````rust
fn test<'a: 'a>(n: bool) -> impl Sized + 'a {
    let true = n else { loop {} };

    loop {}
}


````

original:
````rust
use std::convert::identity;

fn test<'a: 'a>(n: bool) -> impl Sized + 'a {
    //~^ ERROR concrete type differs from previous defining opaque type use
    let true = n else { loop {} };
    let _ = || {
        let _ = identity::<&'a ()>(test(false));
    };
    loop {}
}

fn main() {}

````

Version information
````
rustc 1.79.0-nightly (a385e5667 2024-03-19)
binary: rustc
commit-hash: a385e5667c11098a01368c06e7c0c9236cd477cc
commit-date: 2024-03-19
host: x86_64-unknown-linux-gnu
release: 1.79.0-nightly
LLVM version: 18.1.2

````

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc -Cinstrument-coverage`

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.7IlRFVTMd9Hc/rustc_testrunner_tmpdir_reporting.CuttrTUIRmte/mvce.rs:5:2
  |
5 | }
  |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.7IlRFVTMd9Hc/rustc_testrunner_tmpdir_reporting.CuttrTUIRmte/mvce.rs`

thread 'rustc' panicked at compiler/rustc_mir_transform/src/coverage/counters.rs:125:9:
attempt to add a `Counter` to a BCB target with existing incoming edge counters
stack backtrace:
   0:     0x72da1e807fa5 - std::backtrace_rs::backtrace::libunwind::trace::h663163b1406319df
                               at /rustc/a385e5667c11098a01368c06e7c0c9236cd477cc/library/std/src/../../backtrace/src/backtrace/libunwind.rs:105:5
   1:     0x72da1e807fa5 - std::backtrace_rs::backtrace::trace_unsynchronized::hb98458c0d221a1d2
                               at /rustc/a385e5667c11098a01368c06e7c0c9236cd477cc/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x72da1e807fa5 - std::sys_common::backtrace::_print_fmt::hfd766dc9518113a4
                               at /rustc/a385e5667c11098a01368c06e7c0c9236cd477cc/library/std/src/sys_common/backtrace.rs:68:5
   3:     0x72da1e807fa5 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hc6cd87bcd67d69a9
                               at /rustc/a385e5667c11098a01368c06e7c0c9236cd477cc/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x72da1e858feb - core::fmt::rt::Argument::fmt::hf5b735098c1f998e
                               at /rustc/a385e5667c11098a01368c06e7c0c9236cd477cc/library/core/src/fmt/rt.rs:142:9
   5:     0x72da1e858feb - core::fmt::write::h4b9ff19af3135272
                               at /rustc/a385e5667c11098a01368c06e7c0c9236cd477cc/library/core/src/fmt/mod.rs:1153:17
   6:     0x72da1e7fccbf - std::io::Write::write_fmt::h34d11d08e78b6c42
                               at /rustc/a385e5667c11098a01368c06e7c0c9236cd477cc/library/std/src/io/mod.rs:1843:15
   7:     0x72da1e807d7e - std::sys_common::backtrace::_print::h3bee6832a4022ced
                               at /rustc/a385e5667c11098a01368c06e7c0c9236cd477cc/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x72da1e807d7e - std::sys_common::backtrace::print::hc5073bb2e7688706
                               at /rustc/a385e5667c11098a01368c06e7c0c9236cd477cc/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x72da1e80a9f9 - std::panicking::default_hook::{{closure}}::hf1764864114d23cb
  10:     0x72da1e80a763 - std::panicking::default_hook::h34c466ca2272a795
                               at /rustc/a385e5667c11098a01368c06e7c0c9236cd477cc/library/std/src/panicking.rs:292:9
  11:     0x72da1b3bf0ef - std[806f4cc65fe479e6]::panicking::update_hook::<alloc[ccda606a89b8c1c7]::boxed::Box<rustc_driver_impl[674c6efebf1505bc]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x72da1e80b150 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hc566869933b3317f
                               at /rustc/a385e5667c11098a01368c06e7c0c9236cd477cc/library/alloc/src/boxed.rs:2029:9
  13:     0x72da1e80b150 - std::panicking::rust_panic_with_hook::hbb52b7d77653304f
                               at /rustc/a385e5667c11098a01368c06e7c0c9236cd477cc/library/std/src/panicking.rs:783:13
  14:     0x72da1e80ae6d - std::panicking::begin_panic_handler::{{closure}}::hccee28a87d0405b4
                               at /rustc/a385e5667c11098a01368c06e7c0c9236cd477cc/library/std/src/panicking.rs:649:13
  15:     0x72da1e808469 - std::sys_common::backtrace::__rust_end_short_backtrace::hddc11ac6ccb21feb
                               at /rustc/a385e5667c11098a01368c06e7c0c9236cd477cc/library/std/src/sys_common/backtrace.rs:171:18
  16:     0x72da1e80abd7 - rust_begin_unwind
                               at /rustc/a385e5667c11098a01368c06e7c0c9236cd477cc/library/std/src/panicking.rs:645:5
  17:     0x72da1e855486 - core::panicking::panic_fmt::h0a581737015ab49a
                               at /rustc/a385e5667c11098a01368c06e7c0c9236cd477cc/library/core/src/panicking.rs:72:14
  18:     0x72da1b9e3c76 - <rustc_mir_transform[1307ec64b32b2baf]::coverage::counters::CoverageCounters>::set_bcb_counter
  19:     0x72da1ba1cf6a - <rustc_mir_transform[1307ec64b32b2baf]::coverage::counters::MakeBcbCounters>::get_or_make_counter_operand
  20:     0x72da1b9ebe64 - <rustc_mir_transform[1307ec64b32b2baf]::coverage::InstrumentCoverage as rustc_middle[d868e850dbd6021a]::mir::MirPass>::run_pass
  21:     0x72da1c60198d - rustc_mir_transform[1307ec64b32b2baf]::pass_manager::run_passes_inner
  22:     0x72da1a0211a8 - rustc_mir_transform[1307ec64b32b2baf]::mir_promoted
  23:     0x72da1cad6a9a - rustc_query_impl[c21e40cbcd779a75]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[c21e40cbcd779a75]::query_impl::mir_promoted::dynamic_query::{closure#2}::{closure#0}, rustc_middle[d868e850dbd6021a]::query::erase::Erased<[u8; 16usize]>>
  24:     0x72da1cad6d58 - rustc_query_system[139fc1bc295e0919]::query::plumbing::try_execute_query::<rustc_query_impl[c21e40cbcd779a75]::DynamicConfig<rustc_query_system[139fc1bc295e0919]::query::caches::VecCache<rustc_span[f124091ac140f0eb]::def_id::LocalDefId, rustc_middle[d868e850dbd6021a]::query::erase::Erased<[u8; 16usize]>>, false, false, false>, rustc_query_impl[c21e40cbcd779a75]::plumbing::QueryCtxt, false>
  25:     0x72da1cad69a3 - rustc_query_impl[c21e40cbcd779a75]::query_impl::mir_promoted::get_query_non_incr::__rust_end_short_backtrace
  26:     0x72da1c908072 - rustc_borrowck[1bfd479efec9d870]::mir_borrowck
  27:     0x72da1c907fa7 - rustc_query_impl[c21e40cbcd779a75]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[c21e40cbcd779a75]::query_impl::mir_borrowck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[d868e850dbd6021a]::query::erase::Erased<[u8; 8usize]>>
  28:     0x72da1c909fae - rustc_query_system[139fc1bc295e0919]::query::plumbing::try_execute_query::<rustc_query_impl[c21e40cbcd779a75]::DynamicConfig<rustc_query_system[139fc1bc295e0919]::query::caches::VecCache<rustc_span[f124091ac140f0eb]::def_id::LocalDefId, rustc_middle[d868e850dbd6021a]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[c21e40cbcd779a75]::plumbing::QueryCtxt, false>
  29:     0x72da1c909acc - rustc_query_impl[c21e40cbcd779a75]::query_impl::mir_borrowck::get_query_non_incr::__rust_end_short_backtrace
  30:     0x72da1d0b1d4b - rustc_middle[d868e850dbd6021a]::query::plumbing::query_get_at::<rustc_query_system[139fc1bc295e0919]::query::caches::VecCache<rustc_span[f124091ac140f0eb]::def_id::LocalDefId, rustc_middle[d868e850dbd6021a]::query::erase::Erased<[u8; 8usize]>>>
  31:     0x72da1d0abfad - rustc_hir_analysis[cdf5a959bdcd3d71]::collect::type_of::type_of_opaque
  32:     0x72da1d0abe37 - rustc_query_impl[c21e40cbcd779a75]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[c21e40cbcd779a75]::query_impl::type_of_opaque::dynamic_query::{closure#2}::{closure#0}, rustc_middle[d868e850dbd6021a]::query::erase::Erased<[u8; 8usize]>>
  33:     0x72da1c6f035c - rustc_query_system[139fc1bc295e0919]::query::plumbing::try_execute_query::<rustc_query_impl[c21e40cbcd779a75]::DynamicConfig<rustc_query_system[139fc1bc295e0919]::query::caches::DefIdCache<rustc_middle[d868e850dbd6021a]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[c21e40cbcd779a75]::plumbing::QueryCtxt, false>
  34:     0x72da1d5bbd5e - rustc_query_impl[c21e40cbcd779a75]::query_impl::type_of_opaque::get_query_non_incr::__rust_end_short_backtrace
  35:     0x72da1cca0416 - rustc_middle[d868e850dbd6021a]::query::plumbing::query_get_at::<rustc_query_system[139fc1bc295e0919]::query::caches::DefIdCache<rustc_middle[d868e850dbd6021a]::query::erase::Erased<[u8; 8usize]>>>
  36:     0x72da1a47585a - rustc_hir_analysis[cdf5a959bdcd3d71]::collect::type_of::type_of
  37:     0x72da1c6f1762 - rustc_query_impl[c21e40cbcd779a75]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[c21e40cbcd779a75]::query_impl::type_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[d868e850dbd6021a]::query::erase::Erased<[u8; 8usize]>>
  38:     0x72da1c6f035c - rustc_query_system[139fc1bc295e0919]::query::plumbing::try_execute_query::<rustc_query_impl[c21e40cbcd779a75]::DynamicConfig<rustc_query_system[139fc1bc295e0919]::query::caches::DefIdCache<rustc_middle[d868e850dbd6021a]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[c21e40cbcd779a75]::plumbing::QueryCtxt, false>
  39:     0x72da1c6efe5b - rustc_query_impl[c21e40cbcd779a75]::query_impl::type_of::get_query_non_incr::__rust_end_short_backtrace
  40:     0x72da1cca0416 - rustc_middle[d868e850dbd6021a]::query::plumbing::query_get_at::<rustc_query_system[139fc1bc295e0919]::query::caches::DefIdCache<rustc_middle[d868e850dbd6021a]::query::erase::Erased<[u8; 8usize]>>>
  41:     0x72da1cca6672 - rustc_hir_analysis[cdf5a959bdcd3d71]::check::check::check_item_type
  42:     0x72da1cccc7aa - rustc_hir_analysis[cdf5a959bdcd3d71]::check::wfcheck::check_well_formed
  43:     0x72da1cccb287 - rustc_query_impl[c21e40cbcd779a75]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[c21e40cbcd779a75]::query_impl::check_well_formed::dynamic_query::{closure#2}::{closure#0}, rustc_middle[d868e850dbd6021a]::query::erase::Erased<[u8; 1usize]>>
  44:     0x72da1cccaa00 - rustc_query_system[139fc1bc295e0919]::query::plumbing::try_execute_query::<rustc_query_impl[c21e40cbcd779a75]::DynamicConfig<rustc_query_system[139fc1bc295e0919]::query::caches::VecCache<rustc_hir[3e94e721b9148021]::hir_id::OwnerId, rustc_middle[d868e850dbd6021a]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[c21e40cbcd779a75]::plumbing::QueryCtxt, false>
  45:     0x72da1ccca77f - rustc_query_impl[c21e40cbcd779a75]::query_impl::check_well_formed::get_query_non_incr::__rust_end_short_backtrace
  46:     0x72da1ccc86ba - rustc_hir_analysis[cdf5a959bdcd3d71]::check::wfcheck::check_mod_type_wf
  47:     0x72da1ccc84e1 - rustc_query_impl[c21e40cbcd779a75]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[c21e40cbcd779a75]::query_impl::check_mod_type_wf::dynamic_query::{closure#2}::{closure#0}, rustc_middle[d868e850dbd6021a]::query::erase::Erased<[u8; 1usize]>>
  48:     0x72da1d0e2430 - rustc_query_system[139fc1bc295e0919]::query::plumbing::try_execute_query::<rustc_query_impl[c21e40cbcd779a75]::DynamicConfig<rustc_query_system[139fc1bc295e0919]::query::caches::DefaultCache<rustc_span[f124091ac140f0eb]::def_id::LocalModDefId, rustc_middle[d868e850dbd6021a]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[c21e40cbcd779a75]::plumbing::QueryCtxt, false>
  49:     0x72da1d0e21db - rustc_query_impl[c21e40cbcd779a75]::query_impl::check_mod_type_wf::get_query_non_incr::__rust_end_short_backtrace
  50:     0x72da1c9056cb - rustc_hir_analysis[cdf5a959bdcd3d71]::check_crate
  51:     0x72da1caa2f2a - rustc_interface[c9d53b999d14d195]::passes::analysis
  52:     0x72da1caa2a99 - rustc_query_impl[c21e40cbcd779a75]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[c21e40cbcd779a75]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[d868e850dbd6021a]::query::erase::Erased<[u8; 1usize]>>
  53:     0x72da1d42f503 - rustc_query_system[139fc1bc295e0919]::query::plumbing::try_execute_query::<rustc_query_impl[c21e40cbcd779a75]::DynamicConfig<rustc_query_system[139fc1bc295e0919]::query::caches::SingleCache<rustc_middle[d868e850dbd6021a]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[c21e40cbcd779a75]::plumbing::QueryCtxt, false>
  54:     0x72da1d42f27f - rustc_query_impl[c21e40cbcd779a75]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  55:     0x72da1d285244 - rustc_interface[c9d53b999d14d195]::interface::run_compiler::<core[55495f5072aa6ef1]::result::Result<(), rustc_span[f124091ac140f0eb]::ErrorGuaranteed>, rustc_driver_impl[674c6efebf1505bc]::run_compiler::{closure#0}>::{closure#0}
  56:     0x72da1d38c945 - std[806f4cc65fe479e6]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[c9d53b999d14d195]::util::run_in_thread_with_globals<rustc_interface[c9d53b999d14d195]::util::run_in_thread_pool_with_globals<rustc_interface[c9d53b999d14d195]::interface::run_compiler<core[55495f5072aa6ef1]::result::Result<(), rustc_span[f124091ac140f0eb]::ErrorGuaranteed>, rustc_driver_impl[674c6efebf1505bc]::run_compiler::{closure#0}>::{closure#0}, core[55495f5072aa6ef1]::result::Result<(), rustc_span[f124091ac140f0eb]::ErrorGuaranteed>>::{closure#0}, core[55495f5072aa6ef1]::result::Result<(), rustc_span[f124091ac140f0eb]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[55495f5072aa6ef1]::result::Result<(), rustc_span[f124091ac140f0eb]::ErrorGuaranteed>>
  57:     0x72da1d38c772 - <<std[806f4cc65fe479e6]::thread::Builder>::spawn_unchecked_<rustc_interface[c9d53b999d14d195]::util::run_in_thread_with_globals<rustc_interface[c9d53b999d14d195]::util::run_in_thread_pool_with_globals<rustc_interface[c9d53b999d14d195]::interface::run_compiler<core[55495f5072aa6ef1]::result::Result<(), rustc_span[f124091ac140f0eb]::ErrorGuaranteed>, rustc_driver_impl[674c6efebf1505bc]::run_compiler::{closure#0}>::{closure#0}, core[55495f5072aa6ef1]::result::Result<(), rustc_span[f124091ac140f0eb]::ErrorGuaranteed>>::{closure#0}, core[55495f5072aa6ef1]::result::Result<(), rustc_span[f124091ac140f0eb]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[55495f5072aa6ef1]::result::Result<(), rustc_span[f124091ac140f0eb]::ErrorGuaranteed>>::{closure#1} as core[55495f5072aa6ef1]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  58:     0x72da1e814919 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::ha82e45e83c882ddf
                               at /rustc/a385e5667c11098a01368c06e7c0c9236cd477cc/library/alloc/src/boxed.rs:2015:9
  59:     0x72da1e814919 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h14806e179a475d27
                               at /rustc/a385e5667c11098a01368c06e7c0c9236cd477cc/library/alloc/src/boxed.rs:2015:9
  60:     0x72da1e814919 - std::sys::pal::unix::thread::Thread::new::thread_start::hb8f0207dba0ec3b7
                               at /rustc/a385e5667c11098a01368c06e7c0c9236cd477cc/library/std/src/sys/pal/unix/thread.rs:108:17
  61:     0x72da182a955a - <unknown>
  62:     0x72da18326a3c - <unknown>
  63:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.79.0-nightly (a385e5667 2024-03-19) running on x86_64-unknown-linux-gnu

note: compiler flags: -C instrument-coverage -Z dump-mir-dir=dir

query stack during panic:
#0 [mir_promoted] promoting constants in MIR for `test`
#1 [mir_borrowck] borrow-checking `test`
#2 [type_of_opaque] computing type of opaque `test::{opaque#0}`
#3 [type_of] computing type of `test::{opaque#0}`
#4 [check_well_formed] checking that `test::{opaque#0}` is well-formed
#5 [check_mod_type_wf] checking that types are well-formed in top-level module
#6 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0601`.

```

</p>
</details>

<!--
query stack:
#0 [mir_promoted] promoting constants in MIR for `test`
#1 [mir_borrowck] borrow-checking `test`
#2 [type_of_opaque] computing type of opaque `test::{opaque#0}`
#3 [type_of] computing type of `test::{opaque#0}`
#4 [check_well_formed] checking that `test::{opaque#0}` is well-formed
#5 [check_mod_type_wf] checking that types are well-formed in top-level module
#6 [analysis] running analysis passes on this crate

-->

--- a/compiler/rustc_mir_transform/src/coverage/counters.rs
+++ b/compiler/rustc_mir_transform/src/coverage/counters.rs
@@ -1,7 +1,6 @@
 use rustc_data_structures::captures::Captures;
 use rustc_data_structures::fx::FxHashMap;
 use rustc_data_structures::graph::WithNumNodes;
-use rustc_index::bit_set::BitSet;
 use rustc_index::IndexVec;
 use rustc_middle::mir::coverage::*;
 
@@ -18,10 +17,6 @@ pub(super) enum BcbCounter {
 }
 
 impl BcbCounter {
-    fn is_expression(&self) -> bool {
-        matches!(self, Self::Expression { .. })
-    }
-
     pub(super) fn as_term(&self) -> CovTerm {
         match *self {
             BcbCounter::Counter { id, .. } => CovTerm::Counter(id),
@@ -60,10 +55,6 @@ pub(super) struct CoverageCounters {
     /// We currently don't iterate over this map, but if we do in the future,
     /// switch it back to `FxIndexMap` to avoid query stability hazards.
     bcb_edge_counters: FxHashMap<(BasicCoverageBlock, BasicCoverageBlock), BcbCounter>,
-    /// Tracks which BCBs have a counter associated with some incoming edge.
-    /// Only used by assertions, to verify that BCBs with incoming edge
-    /// counters do not have their own physical counters (expressions are allowed).
-    bcb_has_incoming_edge_counters: BitSet<BasicCoverageBlock>,
     /// Table of expression data, associating each expression ID with its
     /// corresponding operator (+ or -) and its LHS/RHS operands.
     expressions: IndexVec<ExpressionId, Expression>,
@@ -83,7 +74,6 @@ impl CoverageCounters {
             counter_increment_sites: IndexVec::new(),
             bcb_counters: IndexVec::from_elem_n(None, num_bcbs),
             bcb_edge_counters: FxHashMap::default(),
-            bcb_has_incoming_edge_counters: BitSet::new_empty(num_bcbs),
             expressions: IndexVec::new(),
         };
 
@@ -122,14 +112,6 @@ impl CoverageCounters {
     }
 
     fn set_bcb_counter(&mut self, bcb: BasicCoverageBlock, counter_kind: BcbCounter) -> BcbCounter {
-        assert!(
-            // If the BCB has an edge counter (to be injected into a new `BasicBlock`), it can also
-            // have an expression (to be injected into an existing `BasicBlock` represented by this
-            // `BasicCoverageBlock`).
-            counter_kind.is_expression() || !self.bcb_has_incoming_edge_counters.contains(bcb),
-            "attempt to add a `Counter` to a BCB target with existing incoming edge counters"
-        );
-
         if let Some(replaced) = self.bcb_counters[bcb].replace(counter_kind) {
             bug!(
                 "attempt to set a BasicCoverageBlock coverage counter more than once; \
@@ -146,19 +128,6 @@ impl CoverageCounters {
         to_bcb: BasicCoverageBlock,
         counter_kind: BcbCounter,
     ) -> BcbCounter {
-        // If the BCB has an edge counter (to be injected into a new `BasicBlock`), it can also
-        // have an expression (to be injected into an existing `BasicBlock` represented by this
-        // `BasicCoverageBlock`).
-        if let Some(node_counter) = self.bcb_counter(to_bcb)
-            && !node_counter.is_expression()
-        {
-            bug!(
-                "attempt to add an incoming edge counter from {from_bcb:?} \
-                when the target BCB already has {node_counter:?}"
-            );
-        }
-
-        self.bcb_has_incoming_edge_counters.insert(to_bcb);
         if let Some(replaced) = self.bcb_edge_counters.insert((from_bcb, to_bcb), counter_kind) {
             bug!(
                 "attempt to set an edge counter more than once; from_bcb: \

--- a/compiler/rustc_mir_transform/src/coverage/counters.rs
+++ b/compiler/rustc_mir_transform/src/coverage/counters.rs
@@ -1,12 +1,12 @@
+use std::fmt::{self, Debug};
+
 use rustc_data_structures::captures::Captures;
 use rustc_data_structures::fx::FxHashMap;
 use rustc_data_structures::graph::WithNumNodes;
 use rustc_index::IndexVec;
-use rustc_middle::mir::coverage::*;
+use rustc_middle::mir::coverage::{CounterId, CovTerm, Expression, ExpressionId, Op};
 
-use super::graph::{BasicCoverageBlock, CoverageGraph, TraverseCoverageGraphWithLoops};
-
-use std::fmt::{self, Debug};
+use crate::coverage::graph::{BasicCoverageBlock, CoverageGraph, TraverseCoverageGraphWithLoops};
 
 /// The coverage counter or counter expression associated with a particular
 /// BCB node or BCB edge.

--- a/tests/coverage/let_else_loop.cov-map
+++ b/tests/coverage/let_else_loop.cov-map
@@ -1,0 +1,30 @@
+Function name: let_else_loop::_if (unused)
+Raw bytes (19): 0x[01, 01, 00, 03, 00, 16, 01, 01, 0c, 00, 02, 09, 00, 10, 00, 02, 09, 00, 10]
+Number of files: 1
+- file 0 => global file 1
+Number of expressions: 0
+Number of file 0 mappings: 3
+- Code(Zero) at (prev + 22, 1) to (start + 1, 12)
+- Code(Zero) at (prev + 2, 9) to (start + 0, 16)
+- Code(Zero) at (prev + 2, 9) to (start + 0, 16)
+
+Function name: let_else_loop::_loop_either_way (unused)
+Raw bytes (19): 0x[01, 01, 00, 03, 00, 0f, 01, 01, 14, 00, 01, 1c, 00, 23, 00, 01, 05, 00, 0c]
+Number of files: 1
+- file 0 => global file 1
+Number of expressions: 0
+Number of file 0 mappings: 3
+- Code(Zero) at (prev + 15, 1) to (start + 1, 20)
+- Code(Zero) at (prev + 1, 28) to (start + 0, 35)
+- Code(Zero) at (prev + 1, 5) to (start + 0, 12)
+
+Function name: let_else_loop::loopy
+Raw bytes (19): 0x[01, 01, 00, 03, 01, 09, 01, 01, 14, 00, 01, 1c, 00, 23, 05, 01, 01, 00, 02]
+Number of files: 1
+- file 0 => global file 1
+Number of expressions: 0
+Number of file 0 mappings: 3
+- Code(Counter(0)) at (prev + 9, 1) to (start + 1, 20)
+- Code(Zero) at (prev + 1, 28) to (start + 0, 35)
+- Code(Counter(1)) at (prev + 1, 1) to (start + 0, 2)
+

--- a/tests/coverage/let_else_loop.coverage
+++ b/tests/coverage/let_else_loop.coverage
@@ -1,6 +1,6 @@
    LL|       |#![feature(coverage_attribute)]
    LL|       |//@ edition: 2021
-   LL|       |//@ ignore-test
+   LL|       |
    LL|       |// Regression test for <https://github.com/rust-lang/rust/issues/122738>.
    LL|       |// These code patterns should not trigger an ICE when allocating a physical
    LL|       |// counter to a node and also one of its in-edges, because that is allowed

--- a/tests/coverage/let_else_loop.coverage
+++ b/tests/coverage/let_else_loop.coverage
@@ -1,0 +1,35 @@
+   LL|       |#![feature(coverage_attribute)]
+   LL|       |//@ edition: 2021
+   LL|       |//@ ignore-test
+   LL|       |// Regression test for <https://github.com/rust-lang/rust/issues/122738>.
+   LL|       |// These code patterns should not trigger an ICE when allocating a physical
+   LL|       |// counter to a node and also one of its in-edges, because that is allowed
+   LL|       |// when the node contains a tight loop to itself.
+   LL|       |
+   LL|      1|fn loopy(cond: bool) {
+   LL|      1|    let true = cond else { loop {} };
+                                         ^0
+   LL|      1|}
+   LL|       |
+   LL|       |// Variant that also has `loop {}` on the success path.
+   LL|       |// This isn't needed to catch the original ICE, but might help detect regressions.
+   LL|      0|fn _loop_either_way(cond: bool) {
+   LL|      0|    let true = cond else { loop {} };
+   LL|      0|    loop {}
+   LL|       |}
+   LL|       |
+   LL|       |// Variant using regular `if` instead of let-else.
+   LL|       |// This doesn't trigger the original ICE, but might help detect regressions.
+   LL|      0|fn _if(cond: bool) {
+   LL|      0|    if cond {
+   LL|      0|        loop {}
+   LL|       |    } else {
+   LL|      0|        loop {}
+   LL|       |    }
+   LL|       |}
+   LL|       |
+   LL|       |#[coverage(off)]
+   LL|       |fn main() {
+   LL|       |    loopy(true);
+   LL|       |}
+

--- a/tests/coverage/let_else_loop.rs
+++ b/tests/coverage/let_else_loop.rs
@@ -1,0 +1,33 @@
+#![feature(coverage_attribute)]
+//@ edition: 2021
+//@ ignore-test
+// Regression test for <https://github.com/rust-lang/rust/issues/122738>.
+// These code patterns should not trigger an ICE when allocating a physical
+// counter to a node and also one of its in-edges, because that is allowed
+// when the node contains a tight loop to itself.
+
+fn loopy(cond: bool) {
+    let true = cond else { loop {} };
+}
+
+// Variant that also has `loop {}` on the success path.
+// This isn't needed to catch the original ICE, but might help detect regressions.
+fn _loop_either_way(cond: bool) {
+    let true = cond else { loop {} };
+    loop {}
+}
+
+// Variant using regular `if` instead of let-else.
+// This doesn't trigger the original ICE, but might help detect regressions.
+fn _if(cond: bool) {
+    if cond {
+        loop {}
+    } else {
+        loop {}
+    }
+}
+
+#[coverage(off)]
+fn main() {
+    loopy(true);
+}

--- a/tests/coverage/let_else_loop.rs
+++ b/tests/coverage/let_else_loop.rs
@@ -1,6 +1,6 @@
 #![feature(coverage_attribute)]
 //@ edition: 2021
-//@ ignore-test
+
 // Regression test for <https://github.com/rust-lang/rust/issues/122738>.
 // These code patterns should not trigger an ICE when allocating a physical
 // counter to a node and also one of its in-edges, because that is allowed

